### ISSUE TITLE
Portability issue in `xtl_interface.pl` file `abolish_all/1` predicate

### ISSUE THREAD

**Issue by pmoura** [hhu-stups/prob-issues#204]
The `xtl_interface.pl` file defines a local `abolish_all/1` predicate as follows:

```logtalk
:- if(\+ current_prolog_flag(dialect, sicstus)).
abolish_all([]).
abolish_all([Pred|Preds]) :-
    abolish(Pred),
    abolish_all(Preds).
:- else.
abolish_all(Preds) :-
    abolish(Preds, [force(true),tree(true)]).
:- endif.
```

In the calls to this local predicate, all arguments are for dynamic predicates (thankfully; in the ISO Prolog standard the `abolish/1` predicate cannot be used with static predicates). Moreover, there's e legacy `abolish/2` predicate, inherited from Edinburgh Prolog and implemented by several systems (including SWI-Prolog) but with different semantics from SICStus Prolog. Therefore, to improve portability, my suggestion is to change simplify the code above to:

```logtalk
abolish_all([]).
abolish_all([Pred|Preds]) :-
    abolish(Pred),
    abolish_all(Preds).
```

**Comment by dgelessus** [comment]
Hmm. Looking just at the source code, you would be right. The problem is that the XTL module consults external Prolog files that are expected to define their own `prop/2`, `trans/3`, `start/1`, etc. predicates. Those definitions would then be static, because AFAIK `abolish` will remove the previous `dynamic` declaration, and in any case we disable `redefine_warnings` while consulting. I'll have to check, but I think the `abolish(..., [force(true)])` is needed on SICStus to reliably remove the consulted static predicates in this case.

**Comment by pmoura** [comment]
Can't those external Prolog files declare the predicates as dynamic themselves?

**Comment by dgelessus** [comment]
Yes, but those files come from the user, so we have no control over that. I checked our collection of example files, and none of the XTL .P files contain `:- dynamic prop/2`, etc. declarations. So requiring such declarations now would make ProB incompatible with basically all existing XTL files, which would be A Bad Thing™.

In some other places in ProB, we use `load_files(..., [compilation_mode(assert_all)])` to load external files, which apparently forces all predicates to be considered dynamic. That might be the best solution here - that way plain `abolish/1` should always work.

SWI doesn't support the `compilation_mode` option, but its `abolish/1` also works on static predicates, so plain `consult` isn't a problem there. Of course, this a very SICStus/SWI-specific solution, but I think there's no truly portable alternative, without requiring users to change their code...

**Comment by pmoura** [comment]
Fully agree that introducing an incompatibility is not acceptable. In an hypothetical Logtalk port, the (fully portable) solution would be to load the Prolog files using:

```text
?- logtalk_load(..., [hook(add_dynamic_directives)]).
...
```

```logtalk
:- object(add_dynamic_directives,
	implements(expanding)).

	term_expansion(
		begin_of_file
		[	begin_of_file,
			(:- dynamic(prop/2)),
			...
		]
	).

:- end_object.
```

The expansion defined in the `add_dynamic_directives` object is only applied when using that `logtalk_load/2` query `hook/1` option and thus doesn't contaminate anything else.

SWI-Prolog copied recently Logtalk's `begin_of_file` virtual term (after an example I posted in its discussion forum). But I found out (after I added support for this virtual term) that SICStus was first here but using a `beginning_of_file` term (odd choice given the widely supported `end_of_file` virtual term). I guess a solution could be hacked using these virtual terms and conditional compilation directives to approximate the Logtalk solution described above. Still, Prolog implementations of term-expansion (like in SICStus and SWI) are terrible at scoping expansions (see the discussion on hook objects at  https://logtalk.org/manuals/userman/expansion.html if you're curious).

**Comment by leuschel** [comment]
Note: we also have to consider performance. Currently, when running ProB from source, the XTL files can be loaded and compiled as static predicates. This can make a big difference in execution speed. (When running the compiled version of ProB this does not apply; we would need the extended runtime from SICStus to be able to compile user-code from XTL files at runtime.)

**Comment by pmoura** [comment]
Is ProB loading multiple XTL files concurrently? Or is abolishing predicates required when moving from a XTL file to the next XTL file?

**Comment by leuschel** [comment]
It is required when moving from one XTL file to the next, when the user opens a new XTL file (or a B or CSP file)
